README.md
service.subtitles.ktuvit
==================

Ktuvit.me subtitle service for Kodi

In order to get the encoded password:

1. Open [ktuvit.me](https://www.ktuvit.me) in the browser
2. Open [developer tools](https://developers.google.com/web/tools/chrome-devtools/open)
3. Enter this code in the **console**: 
```javascript
x = { value: 'MY-APSSWORD' };
loginHandler.EncryptPassword({}, x, 'MY@EMAIL.COM');
copy(x.value); // this will copy your encrtyped password to your clipboard
console.log(`Now past it in the addon's settings at the Encrypted password field`)
``` 
![Gujn9Y8vTF](https://user-images.githubusercontent.com/9304194/94992868-897f0100-0595-11eb-8694-0272ae2f19b9.gif)



changelog.txt
2.2.1 - by CaTz (03/10/2020)
- Downgrade beautifulsoup version

2.2.0 - by CaTz (03/10/2020)
- Add movie subs download
- Add login ability in addon's settings

2.1.0 - by CaTz (03/10/2020)
- Filter better tv show search results
- Add language flag to subs list

2.0.0 - by CaTz (30/09/2020)
- Rename to ktuvit.me
- Add support for downloading tvshow subs

1.2.0 - by sagiben (13/12/2019)
- Add support for KODI 19

1.1.1 - by CaTz (01/12/2019)
- bug-fix: write on the correct dict

1.1.0 - by CaTz (30/11/2019)
- Add the ability to search subs on focused item.

1.0.0 - by CaTz (23/11/2019)
- Initial release of the addon



tests/stubs/xbmcaddon.py
## @package xbmcaddon
#  A class to access addon properties.
#

#noinspection PyUnusedLocal
class Addon(object):
    def __init__(self, id=None):
        """Creates a new Addon class.

        id: string - id of the addon (autodetected in XBMC Eden)

        Example:
            self.Addon = xbmcaddon.Addon(id='script.recentlyadded')
        """
        pass

    def getLocalizedString(self, id):
        """Returns an addon's localized 'unicode string'.

        id: integer - id# for string you want to localize.

        Example:
            locstr = self.Addon.getLocalizedString(id=6)
        """
        return id

    def getSetting(self, id):
        return unicode

    def setSetting(self, id, value):
        """Sets a script setting.

        id: string - id of the setting that the module needs to access.
        value: string or unicode - value of the setting.

        Example:
            self.Settings.setSetting(id='username', value='teamxbmc')
        """
        pass

    def openSettings(self):
        """Opens this scripts settings dialog."""
        pass

    def getAddonInfo(self, id):
        """Returns the value of an addon property as a string.

        id: string - id of the property that the module needs to access.

        Note:
            Choices are (author, changelog, description, disclaimer, fanart, icon, id, name, path
            profile, stars, summary, type, version)

        Example:
            version = self.Addon.getAddonInfo('version')
        """
        data = {"version": "1.0.0", "name": "ScrewZira",
                "profile": "C:\\Users\\CaTz\\AppData\\Roaming\\XBMC\\userdata\\addon_data\\service.subtitles.subscenter"}
        return data[id]


tests/test_subsUtilities.py
# -*- coding: utf-8 -*-
import test_helpers
from unittest import TestCase
from resources.lib.SUBUtilities import SubsHelper, parse_rls_title


class TestSUBUtilities(TestCase):
    def setUp(self):
        self.helper = SubsHelper()

    def test_get_subtitle_list(self):
        item = {'episode': '1', 'title': 'The Blood of Man', 'season': '2', 'year': '', 'tvshow': "Da Vinci's Demons",
                '3let_language': ['eng', 'heb']}

        result = self.helper._search(item)
        self.assertEqual(result[0]['name'], item["tvshow"])
        self.assertGreater(len(result[0]['subs']['he']), 0)

    def test_get_subtitle_list2(self):
        item = {'episode': '', 'title': 'Two.and.a.Half.Men.S11E13.480p.HDTV.X264-DIMENSION',
                'season': '', 'year': '', 'tvshow': '', '3let_language': ['en', 'he']}

        parse_rls_title(item)
        result = self.helper._search(item)
        self.assertEqual(result[0]['name'], 'Two and a Half Men')
        self.assertGreater(len(result[0]['subs']['he']), 0)

    def test_get_subtitle_list3(self):
        item = {'episode': '', 'title': 'Inception', 'season': '', 'year': '2010',
                'tvshow': '', '3let_language': ['en', 'he']}

        result = self.helper._search(item)
        self.assertEqual(result[0]['name'], item['title'])
        self.assertGreater(len(result[0]['subs']['he']), 0)

    def test_get_subtitle_list4(self):
        item = {'episode': '', 'title': 'The.Flash.2014.S02E05.480p.HDTV.X264-DIMENSION.mkv',
                'preferredlanguage': 'heb', 'season': '', 'year': '', 'tvshow': '', '3let_language': ['heb']}

        parse_rls_title(item)
        result = self.helper._search(item)
        self.assertEqual(result[0]['name'], 'The Flash')
        self.assertGreater(len(result[0]['subs']['he']), 0)

    def test_get_subtitle_list5_should_ignore_year_on_tvshow(self):
        item = {'episode': '5', 'title': 'Episode 5', 'preferredlanguage': '', 'season': '3',
                'year': '2016', 'tvshow': 'The Affair', '3let_language': ['eng', 'heb']}

        parse_rls_title(item)
        result = self.helper._search(item)

        self.assertEqual(result[0]['name'], item["tvshow"])
        self.assertGreater(len(result[0]['subs']['he']), 0)


service.py
# -*- coding: utf-8 -*-

import os
import sys

import xbmc
import xbmcaddon
import xbmcgui
import xbmcplugin
import xbmcvfs

try:
    # Python 3 - Kodi 19
    from urllib.parse import urlparse, unquote
    from urllib.request import Request
except ImportError:
    # Python 2 - Kodi 18 and below
    from urlparse import urlparse, unquote
    from urllib2 import Request

__addon__ = xbmcaddon.Addon()
__author__ = __addon__.getAddonInfo('author')
__scriptid__ = __addon__.getAddonInfo('id')
__scriptname__ = __addon__.getAddonInfo('name')
__version__ = __addon__.getAddonInfo('version')
__language__ = __addon__.getLocalizedString

__cwd__ = xbmc.translatePath(__addon__.getAddonInfo('path'))
__profile__ = xbmc.translatePath(__addon__.getAddonInfo('profile'))
__resource__ = xbmc.translatePath(os.path.join(__cwd__, 'resources', 'lib'))
__temp__ = xbmc.translatePath(os.path.join(__profile__, 'temp'))

sys.path.append(__resource__)

from SUBUtilities import SubsHelper, log, normalizeString, parse_rls_title, clean_title


def search(item):
    helper = SubsHelper()
    subtitles_list = helper.get_subtitle_list(item)
    if subtitles_list:
        for it in subtitles_list:
            listitem = xbmcgui.ListItem(label=it["language_name"], label2=it["filename"])
            listitem.setArt({'icon': str(it["rating"]), 'thumb': it["language_flag"]})

            if it["sync"]:
                listitem.setProperty("sync", "true")
            else:
                listitem.setProperty("sync", "false")

            if it.get("hearing_imp", False):
                listitem.setProperty("hearing_imp", "true")
            else:
                listitem.setProperty("hearing_imp", "false")

            url = "plugin://%s/?action=download&id=%s&sub_id=%s&filename=%s&language=%s" % (
                __scriptid__, it["id"], it["sub_id"], it["filename"], it["language_flag"])
            xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]), url=url, listitem=listitem, isFolder=False)


def download(id, sub_id, filename, language):
    subtitle_list = []
    exts = [".srt", ".sub"]

    filename = os.path.join(__temp__, "%s.%s.srt" % (filename, language))

    helper = SubsHelper()
    helper.download(id, sub_id, filename)

    for file in xbmcvfs.listdir(__temp__)[1]:
        full_path = os.path.join(__temp__, file)
        if os.path.splitext(full_path)[1] in exts:
            subtitle_list.append(full_path)

    return subtitle_list


def get_params(string=""):
    param = []
    if string == "":
        paramstring = sys.argv[2]
    else:
        paramstring = string

    if len(paramstring) >= 2:
        params = paramstring
        cleanedparams = params.replace('?', '')
        if (params[len(params) - 1] == '/'):
            params = params[0:len(params) - 2]
        pairsofparams = cleanedparams.split('&')
        param = {}
        for i in range(len(pairsofparams)):
            splitparams = {}
            splitparams = pairsofparams[i].split('=')
            if (len(splitparams)) == 2:
                param[splitparams[0]] = splitparams[1]

    return param


def title_from_focused_item(item_data):
    label_type = xbmc.getInfoLabel("ListItem.DBTYPE")  # movie/tvshow/season/episode
    label_movie_title = xbmc.getInfoLabel("ListItem.OriginalTitle")
    is_movie = xbmc.getCondVisibility("Container.Content(movies)") or label_type == 'movie'
    is_episode = xbmc.getCondVisibility("Container.Content(episodes)") or label_type == 'episode'

    title = ''
    if is_movie and label_movie_title and item_data['year']:
        title = label_movie_title + " " + item_data['year']
    elif is_episode and item_data['tvshow'] and item_data['season'] and item_data['episode']:
        title = ("%s S%.2dE%.2d" % (item_data['tvshow'], int(item_data['season']), int(item_data['episode'])))

    return title


params = get_params()


def collect_initial_data():
    item_data = {
        'temp': False,
        'rar': False,
        '3let_language': [],
        'preferredlanguage': unquote(params.get('preferredlanguage', ''))
    }

    item_data['preferredlanguage'] = xbmc.convertLanguage(item_data['preferredlanguage'], xbmc.ISO_639_2)

    if xbmc.Player().isPlaying():
        item_data['year'] = xbmc.getInfoLabel("VideoPlayer.Year")  # Year
        item_data['season'] = str(xbmc.getInfoLabel("VideoPlayer.Season"))  # Season
        item_data['episode'] = str(xbmc.getInfoLabel("VideoPlayer.Episode"))  # Episode
        item_data['tvshow'] = normalizeString(xbmc.getInfoLabel("VideoPlayer.TVshowtitle"))  # Show
        item_data['title'] = normalizeString(
            xbmc.getInfoLabel("VideoPlayer.OriginalTitle"))  # try to get original title
        item_data['file_original_path'] = unquote(xbmc.Player().getPlayingFile())  # Full path of a playing file

        if item_data['title'] == "":
            log("VideoPlayer.OriginalTitle not found")
            item_data['title'] = normalizeString(
                xbmc.getInfoLabel("VideoPlayer.Title"))  # no original title, get just Title

    else:
        item_data['year'] = xbmc.getInfoLabel("ListItem.Year")
        item_data['season'] = xbmc.getInfoLabel("ListItem.Season")
        item_data['episode'] = xbmc.getInfoLabel("ListItem.Episode")
        item_data['tvshow'] = xbmc.getInfoLabel("ListItem.TVShowTitle")
        item_data['title'] = title_from_focused_item(item_data)
        item_data['file_original_path'] = ""

    return item_data


if params['action'] in ['search', 'manualsearch']:
    log("Version: '%s'" % (__version__,))
    log("Action '%s' called" % (params['action']))

    if params['action'] == 'manualsearch':
        params['searchstring'] = unquote(params['searchstring'])

    item = collect_initial_data()

    if params['action'] == 'manualsearch':
        if item['season'] != '' or item['episode']:
            item['tvshow'] = params['searchstring']
        else:
            item['title'] = params['searchstring']

    for lang in unquote(params['languages']).split(","):
        item['3let_language'].append(xbmc.convertLanguage(lang, xbmc.ISO_639_2))

    log("Item before cleaning: \n    %s" % item)

    # clean title + tvshow params
    clean_title(item)
    parse_rls_title(item)

    if item['episode'].lower().find("s") > -1:  # Check if season is "Special"
        item['season'] = "0"
        item['episode'] = item['episode'][-1:]

    if item['file_original_path'].find("http") > -1:
        item['temp'] = True

    elif item['file_original_path'].find("rar://") > -1:
        item['rar'] = True
        item['file_original_path'] = os.path.dirname(item['file_original_path'][6:])

    elif item['file_original_path'].find("stack://") > -1:
        stackPath = item['file_original_path'].split(" , ")
        item['file_original_path'] = stackPath[0][8:]
    log("%s" % item)
    search(item)


elif params['action'] == 'download':
    ## we pickup all our arguments sent from def search()
    subs = download(params["id"], params["sub_id"], params["filename"], params["language"])
    ## we can return more than one subtitle for multi CD versions, for now we are still working out how to handle that in XBMC core
    for sub in subs:
        listitem = xbmcgui.ListItem(label=sub)
        xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]), url=sub, listitem=listitem, isFolder=False)

elif params['action'] == 'login':
    helper = SubsHelper()
    helper.login(True)
    __addon__.openSettings()

xbmcplugin.endOfDirectory(int(sys.argv[1]))  ## send end of directory to XBMC


resources/lib/SUBUtilities.py
# -*- coding: utf-8 -*-
import os
import re
import unicodedata
import json
import zlib
import shutil

import xbmc
import xbmcvfs
import xbmcaddon
from bs4 import BeautifulSoup

try:
    # Python 3 - Kodi 19
    from http.cookiejar import LWPCookieJar
    from urllib.request import Request, build_opener, HTTPCookieProcessor
    from urllib.parse import urlencode
except ImportError:
    # Python 2 - Kodi 18 and below
    from cookielib import LWPCookieJar
    from urllib2 import Request, build_opener, HTTPCookieProcessor
    from urllib import urlencode

__addon__ = xbmcaddon.Addon()
__version__ = __addon__.getAddonInfo('version')  # Module version
__scriptname__ = __addon__.getAddonInfo('name')
__language__ = __addon__.getLocalizedString
__profile__ = xbmc.translatePath(__addon__.getAddonInfo('profile'))
__temp__ = xbmc.translatePath(os.path.join(__profile__, 'temp', ''))
__kodi_version__ = xbmc.getInfoLabel('System.BuildVersion').split(' ')[0]

regexHelper = re.compile('\W+', re.UNICODE)


# ===============================================================================
# Private utility functions
# ===============================================================================
def normalizeString(_str):
    if not isinstance(_str, str):
        _str = unicodedata.normalize('NFKD', _str)  # .encode('utf-8', 'ignore')
    return _str


def clean_title(item):
    title = os.path.splitext(os.path.basename(item["title"]))
    tvshow = os.path.splitext(os.path.basename(item["tvshow"]))

    if len(title) > 1:
        if re.match(r'^\.[a-z]{2,4}$', title[1], re.IGNORECASE):
            item["title"] = title[0]
        else:
            item["title"] = ''.join(title)
    else:
        item["title"] = title[0]

    if len(tvshow) > 1:
        if re.match(r'^\.[a-z]{2,4}$', tvshow[1], re.IGNORECASE):
            item["tvshow"] = tvshow[0]
        else:
            item["tvshow"] = ''.join(tvshow)
    else:
        item["tvshow"] = tvshow[0]

    # Removes country identifier at the end
    item['title'] = re.sub(r'\([^\)]+\)\W*$', '', item['title']).strip()
    item['tvshow'] = re.sub(r'\([^\)]+\)\W*$', '', item['tvshow']).strip()


def parse_rls_title(item):
    title = regexHelper.sub(' ', item["title"])
    tvshow = regexHelper.sub(' ', item["tvshow"])

    groups = re.findall(r"(.*?) (\d{4})? ?(?:s|season|)(\d{1,2})(?:e|episode|x|\n)(\d{1,2})", title, re.I)

    if len(groups) == 0:
        groups = re.findall(r"(.*?) (\d{4})? ?(?:s|season|)(\d{1,2})(?:e|episode|x|\n)(\d{1,2})", tvshow, re.I)

    if len(groups) > 0 and len(groups[0]) >= 3:
        title, year, season, episode = groups[0]
        item["year"] = str(int(year)) if len(year) == 4 else year

        item["tvshow"] = regexHelper.sub(' ', title).strip()
        item["season"] = str(int(season))
        item["episode"] = str(int(episode))
        log("TV Parsed Item: %s" % (item,))

    else:
        groups = re.findall(r"(.*?)(\d{4})", item["title"], re.I)
        if len(groups) > 0 and len(groups[0]) >= 1:
            title = groups[0][0]
            item["title"] = regexHelper.sub(' ', title).strip()
            item["year"] = groups[0][1] if len(groups[0]) == 2 else item["year"]

            log("MOVIE Parsed Item: %s" % (item,))


def log(msg):
    xbmc.log("### [%s] - %s" % (__scriptname__, msg), level=xbmc.LOGDEBUG)


def notify(msg_id):
    xbmc.executebuiltin((u'Notification(%s,%s)' % (__scriptname__, __language__(msg_id))))


class SubsHelper:
    BASE_URL = "https://www.ktuvit.me/Services"

    def __init__(self):
        self.urlHandler = URLHandler()

    def get_subtitle_list(self, item):
        search_results = self._search(item)
        results = self._build_subtitle_list(search_results, item)

        return results

    # return list of movies / tv-series from the site`s search
    def _search(self, item):
        search_string = re.split(r'\s\(\w+\)$', item["tvshow"])[0] if item["tvshow"] else item["title"]
        log("search_string: %s" % search_string)

        query = {"FilmName": search_string,
                 "Actors": [],
                 "Studios": None,
                 "Directors": [],
                 "Genres": [],
                 "Countries": [],
                 "Languages": [],
                 "Year": "",
                 "Rating": [],
                 "Page": 1,
                 "SearchType": "0",
                 "WithSubsOnly": False}
        if item["tvshow"]:
            query["SearchType"] = "1"
        elif item["year"]:
            query["Year"] = item["year"]

        search_result = self.urlHandler.request(self.BASE_URL + "/ContentProvider.svc/SearchPage_search",
                                                data={"request": query})

        results = []
        log("Results: %s" % search_result)

        if search_result is None or len(search_result["Films"]) == 0:
            notify(32001)
            return results  # return empty set

        ids = self._get_filtered_ids(search_result["Films"], search_string)
        log("Filtered Ids: %s" % ids)

        if item["tvshow"]:
            results = self._search_tvshow(item, ids)
        else:
            results = self._search_movie(ids)

        log("Subtitles: %s" % results)

        return results

    def _search_tvshow(self, item, ids):
        subs = []

        for id in ids:
            query_string = {
                "moduleName": "SubtitlesList",
                "SeriesID": id,
                "Season": item['season'],
                "Episode": item['episode']
            }
            raw_html = self.urlHandler.request(self.BASE_URL + "/GetModuleAjax.ashx", query_string=query_string)

            sub_list = BeautifulSoup(raw_html, "html.parser")
            sub_rows = sub_list.find_all("tr")

            for row in sub_rows:
                columns = row.find_all("td")
                sub = {
                    'id': id
                }
                for index, column in enumerate(columns):
                    if index == 0:
                        sub['rls'] = column.get_text().strip().split("\n")[0]
                    if index == 4:
                        sub['downloads'] = int(column.get_text().strip())
                    if index == 5:
                        sub['sub_id'] = column.find("input", attrs={"data-sub-id": True})["data-sub-id"]

                subs.append(sub)

        return subs

    def _search_movie(self, ids):
        subs = []

        for movie_id in ids:
            query_string = {
                "ID": movie_id,
            }
            raw_html = self.urlHandler.request(self.BASE_URL + "/../MovieInfo.aspx", query_string=query_string)
            html = BeautifulSoup(raw_html, "html.parser")
            sub_rows = html.select("table#subtitlesList tbody > tr")
            log('html %s' % sub_rows)

            for row in sub_rows:
                columns = row.find_all("td")
                sub = {
                    'id': movie_id
                }
                for index, column in enumerate(columns):
                    if index == 0:
                        sub['rls'] = column.get_text().strip().split("\n")[0]
                    if index == 4:
                        sub['downloads'] = int(column.get_text().strip())
                    if index == 5:
                        sub['sub_id'] = column.find("a", attrs={"data-subtitle-id": True})["data-subtitle-id"]

                subs.append(sub)

        return subs

    def _build_subtitle_list(self, search_results, item):
        language = 'he'
        lang3 = xbmc.convertLanguage(language, xbmc.ISO_639_2)
        total_downloads = 0
        ret = []
        for result in search_results:
            title = result["rls"]
            subtitle_rate = self._calc_rating(title, item["file_original_path"])
            total_downloads += result['downloads']

            ret.append({
                'lang_index': item["3let_language"].index(lang3),
                'filename': title,
                'language_name': xbmc.convertLanguage(language, xbmc.ENGLISH_NAME),
                'language_flag': language,
                'id': result["id"],
                'sub_id': result["sub_id"],
                'rating': result['downloads'],
                'sync': subtitle_rate >= 3.8,
                'hearing_imp': False,
                'is_preferred': lang3 == item['preferredlanguage']
            })

        # Fix the rating
        if total_downloads:
            for it in ret:
                log('rating %s totals %s' % (it['rating'], total_downloads))
                it["rating"] = str(int(round(it["rating"] / float(total_downloads), 1) * 5))

        return sorted(ret, key=lambda x: (x['is_preferred'], x['lang_index'], x['sync'], x['rating']), reverse=True)

    def _calc_rating(self, subsfile, file_original_path):
        file_name = os.path.basename(file_original_path)
        folder_name = os.path.split(os.path.dirname(file_original_path))[-1]

        subsfile = re.sub(r'\W+', '.', subsfile).lower()
        file_name = re.sub(r'\W+', '.', file_name).lower()
        folder_name = re.sub(r'\W+', '.', folder_name).lower()
        log("# Comparing Releases:\n [subtitle-rls] %s \n [filename-rls] %s \n [folder-rls] %s" % (
            subsfile, file_name, folder_name))

        subsfile = subsfile.split('.')
        file_name = file_name.split('.')[:-1]
        folder_name = folder_name.split('.')

        if len(file_name) > len(folder_name):
            diff_file = list(set(file_name) - set(subsfile))
            rating = (1 - (len(diff_file) / float(len(file_name)))) * 5
        else:
            diff_folder = list(set(folder_name) - set(subsfile))
            rating = (1 - (len(diff_folder) / float(len(folder_name)))) * 5

        log("\n rating: %f (by %s)" % (round(rating, 1), "file" if len(file_name) > len(folder_name) else "folder"))

        return round(rating, 1)

    def download(self, id, sub_id, filename):
        ## Cleanup temp dir, we recomend you download/unzip your subs in temp folder and
        ## pass that to XBMC to copy and activate
        if xbmcvfs.exists(__temp__):
            shutil.rmtree(__temp__)
        xbmcvfs.mkdirs(__temp__)

        query = {
            "request": {
                "FilmID": id,
                "SubtitleID": sub_id,
                "FontSize": 0,
                "FontColor": "",
                "PredefinedLayout": -1}}

        response = self.urlHandler.request(self.BASE_URL + "/ContentProvider.svc/RequestSubtitleDownload", data=query)
        f = self.urlHandler.request(self.BASE_URL + '/DownloadFile.ashx',
                                    query_string={"DownloadIdentifier": response["DownloadIdentifier"]})
        with open(filename, "wb") as subFile:
            subFile.write(f)
        subFile.close()

    def login(self, notify_success=False):
        email = __addon__.getSetting("email")
        password = __addon__.getSetting("password")
        post_data = {"request": {"Email": email, "Password": password}}

        response = self.urlHandler.request(self.BASE_URL + "/MembershipService.svc/Login", data=post_data)

        if response["IsSuccess"] is True:
            self.urlHandler.save_cookie()
            if notify_success:
                notify(32007)
            return True
        else:
            notify(32005)
            return None

    def _get_filtered_ids(self, list, search_string):
        ids = []

        search_string = regexHelper.sub('', search_string).lower()

        for result in list:
            eng_name = regexHelper.sub('', regexHelper.sub(' ', result['EngName'])).lower()
            heb_name = regexHelper.sub('', result['HebName'])

            if (search_string.startswith(eng_name) or eng_name.startswith(search_string) or
                    search_string.startswith(heb_name) or heb_name.startswith(search_string)):
                ids.append(result["ID"])

        return ids


class URLHandler:
    def __init__(self):
        cookie_filename = os.path.join(__profile__, "cookiejar.txt")
        self.cookie_jar = LWPCookieJar(cookie_filename)
        if os.access(cookie_filename, os.F_OK):
            self.cookie_jar.load()

        self.opener = build_opener(HTTPCookieProcessor(self.cookie_jar))
        self.opener.addheaders = [('Accept-Encoding', 'gzip'),
                                  ('Accept-Language', 'en-us,en;q=0.5'),
                                  ('Pragma', 'no-cache'),
                                  ('Cache-Control', 'no-cache'),
                                  ('Content-type', 'application/json'),
                                  ('User-Agent',
                                   'Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Kodi/%s Chrome/78.0.3904.97 Safari/537.36' % (
                                       __kodi_version__))]

    def request(self, url, data=None, query_string=None, referrer=None, cookie=None):
        if data is not None:
            data = json.dumps(data).encode('utf8')
        if query_string is not None:
            url += '?' + urlencode(query_string)
        if referrer is not None:
            self.opener.addheaders += [('Referrer', referrer)]
        if cookie is not None:
            self.opener.addheaders += [('Cookie', cookie)]

        content = None
        log("Getting url: %s" % (url))
        if data is not None:
            log("Post Data: %s" % (data))
        try:
            req = Request(url, data, headers={'Content-Type': 'application/json'})
            response = self.opener.open(req)
            content = None if response.code != 200 else response.read()

            if response.headers.get('content-encoding', '') == 'gzip':
                try:
                    content = zlib.decompress(content, 16 + zlib.MAX_WBITS)
                except zlib.error:
                    pass

            if response.headers.get('content-type', '').startswith('application/json'):
                parsed_content = json.loads(content, encoding="utf-8")
                content = json.loads(parsed_content["d"], encoding="utf-8")

            response.close()
        except Exception as e:
            log("Failed to get url: %s\n%s" % (url, e))
            # Second parameter is the filename
        return content

    def save_cookie(self):
        # extend cookie expiration
        for cookie in self.cookie_jar:
            if cookie.expires is not None:
                cookie.expires += 2 * 12 * 30 * 24 * 60 * 60

        self.cookie_jar.save()
